data_types/src/database_rules.rs

```diff
@@ -800,6 +800,147 @@ impl PartialEq for Matcher {
 }
 impl Eq for Matcher {}
 
+impl From<ShardConfig> for management::ShardConfig {
+    fn from(shard_config: ShardConfig) -> Self {
+        Self {
+            specific_targets: shard_config.specific_targets.map(|i| i.into()),
+            hash_ring: shard_config.hash_ring.map(|i| i.into()),
+            ignore_errors: shard_config.ignore_errors,
+        }
+    }
+}
+
+impl TryFrom<management::ShardConfig> for ShardConfig {
+    type Error = FieldViolation;
+
+    fn try_from(proto: management::ShardConfig) -> Result<Self, Self::Error> {
+        Ok(Self {
+            specific_targets: proto
+                .specific_targets
+                .map(|i| i.try_into())
+                .map_or(Ok(None), |r| r.map(Some))?,
```

Comment:
To check my own understanding, is this the same as 

```suggestion
                .map(|i| i.try_into())?
```
 
I may be missing some subtlety

Reply:
No, the code you suggested there would try to return a `None` value from the whole function. It doesn't compile:

```
error[E0277]: `?` couldn't convert the error to `generated_types::google::FieldViolation`
   --> data_types/src/database_rules.rs:820:39
    |
820 |                 .map(|i| i.try_into())?,
    |                                       ^ the trait `From<std::option::NoneError>` is not implemented for `generated_types::google::FieldViolation`
```

(and isn't the behavior you want)

Reply:
I learned this pattern (of turning an `Option<Result<...>>` into a `Result<Option<...`) from https://doc.rust-lang.org/beta/rust-by-example/error/multiple_error_types/option_result.html

Reply:
I think you maybe can use https://doc.rust-lang.org/std/option/enum.Option.html#method.transpose for a similar purpose

Reply:
Wow rust can have methods on specialization impls. I'm impressed.